dask/dataframe/io/parquet/arrow.py

```diff
@@ -821,8 +837,50 @@ def initialize_write(
         partition_on=None,
         ignore_divisions=False,
         division_info=None,
+        schema=None,
+        index_cols=None,
         **kwargs,
     ):
+        # Infer schema if "infer"
+        # (also start with inferred schema if user passes a dict)
+        if schema == "infer" or isinstance(schema, dict):
+
+            # Start with schema from _meta_nonempty
+            _schema = pa.Schema.from_pandas(
+                df._meta_nonempty.set_index(index_cols)
+                if index_cols
+                else df._meta_nonempty
+            )
+
+            # Use dict to update our inferred schema
+            if isinstance(schema, dict):
+                schema = pa.schema(schema)
+                for name in schema.names:
+                    i = _schema.get_field_index(name)
+                    j = schema.get_field_index(name)
+                    _schema = _schema.set(i, schema.field(j))
+
+            # If we have object columns, we need to sample partitions
+            # until we find non-null data for each column in `sample`
+            sample = [col for col in df.columns if df[col].dtype == "object"]
+            if schema_field_supported and sample and schema == "infer":
+                for i in range(df.npartitions):
+                    # Keep data on worker
+                    _s = delayed(lambda x: pa.Schema.from_pandas(x))(
+                        df[sample].to_delayed()[i]
+                    ).compute()
```

Comment:
I wonder does it seem possible that some pathological cases might involve enough partitions that re-building the delayed func and delayed partitions in every iteration of this for loop would be doing a lot of unnecessary work?  Would it make sense to do something like this?
```python
# One time delayed func def at module top level or at top of function body
d_schema_from_pandas = delayed(pa.Schema.from_pandas)

# One time construction of delayed partitions
d_partitions = df[sample].to_delayed()
```
Then the line inside the for loop could look like this:
```suggestion
                    _s = d_schema_from_pandas(d_partitions[i])
```
Although maybe `df[sample].to_delayed()` could still be good to have in the for loop since it would compute fewer and fewer columns as the iterations continue.  Not sure about that.

Reply:
`df[sample].to_delayed()[i]` should be roughly free, but maybe worth checking. Is that what you mean? 
It might be cleaner to call just once, as you say, but I doubt it makes a performance difference.

Reply:
> df[sample].to_delayed()[i] should be roughly free, but maybe worth checking. Is that what you mean?

Yeah, I just wasn't sure if we could get away with doing it outside of the for loop.  But I guess that loop should never run many times so it probably doesn't matter like you say.  Also, the work being done per iteration is a lot more than the work to evaluate `df[sample].to_delayed[i]`.  So yeah, never mind about this too heh.

Although at least I think you can use `delayed(pa.Schema.from_pandas)` with no lambda since `from_pandas` is already a unary func.

Reply:
> I think you can use delayed(pa.Schema.from_pandas)

That should be true, and a bit neater

Reply:
@davesque - Thanks for the suggestion.  I removed the lambda in favor of a one-time `delayed(pa.Schema.from_pandas)` operation. 

Regarding the `df[sample].to_delayed[i]`: I certainly wanted to avoid doing this every loop, but realized the "pathological" case (where we need to sample many partitions) should probably not be the focus.  I think the most common scenario will require a few partitions (at most), and the benefit of narrowing the columns in `sample` may outweigh the cost of the extra `to_delayed` call.  If the user is expecting a pathological case, they can always pass in the schema for "object" columns (or all columns) directly... That said, I'm happy to reference a single `d_partitions` object if you/others feel that it makes the code more readable (or is indeed faster).

Reply:
> I think the most common scenario will require a few partitions (at most), and the benefit of narrowing the columns in sample may outweigh the cost of the extra to_delayed call.

I think you're right about this.  The column narrowing will probably always outweigh the savings of making the delayed objects.

> I'm happy to reference a single d_partitions object if you/others feel that it makes the code more readable (or is indeed faster).

Nah, I think with the removal of the lambda thing it all looks good.  Thanks for humoring me on all this.